### PR TITLE
Support up to Openstack Queens

### PR DESCRIPTION
**Pull request by jgagnaire**

### CODE REVIEW

File: swift_scality_backend/diskfile.py [hunk 1]
```diff
@@ -277,6 +293,31 @@ def __repr__(self):
...
+        include any persistent metadata that was set by the original PUT.
+        """
+        md_dest = {}
+
+        if not mf_md_source:
+            md_dest.update(df_md_source)
+        else:
+            sys_metadata = {
+                key: val for key, val in df_md_source.items()
+                if key in (RESERVED_DATAFILE_META | DATAFILE_SYSTEM_META) or
```
[the remark this answers is not in the document]
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: Fixed

File: swift_scality_backend/diskfile.py [hunk 2]
```diff
@@ -109,11 +118,20 @@ def put(self, metadata):
...
 
         self._release_conn()
-        metadata['name'] = self._name
+
+        metadata_to_put = copy.copy(metadata)
+        metadata_to_put['name'] = self._name
+        metadata_to_put['ETag'] = self._md5sum.hexdigest()
+
+        metadata_to_put.update({
+            'df': metadata,
```
Comment (NicolasT) [inline swift_scality_backend/diskfile.py]: Are `df` and `mf` names we choose, or enforced by Swift?

File: swift_scality_backend/diskfile.py [hunk 3]
```diff
@@ -277,6 +295,31 @@ def __repr__(self):
...
+
+        Metafile metadata refers to metadata written by a POST, and does not
+        include any persistent metadata that was set by the original PUT.
+        """
+        md_dest = {}
+
+        if not mf_md_source:
+            md_dest.update(df_md_source)
+        else:
+            sys_metadata = {
```
Comment (NicolasT) [inline swift_scality_backend/diskfile.py]: This breaks Python 2.6 compatibility (but maybe we no longer support that anyway?)

File: swift_scality_backend/diskfile.py [hunk 4]
```diff
@@ -354,12 +403,39 @@ def create(self, size=None):
...
+
+        if not self._metadata:
+            md_to_write = copy.deepcopy(md_to_add)
+            md_to_write.update({
+                'df': {
+                    'name': self._name,
+                },
+                'mf': md_to_add
+            })
+            self.client_collection.get_write_client().put_meta(self._name, md_to_write)
```
Comment (NicolasT) [inline swift_scality_backend/diskfile.py]: Above there are some changes which turn `self.client_collection` into `self._client_collection`, and here it's `self.client_collection` again. A bit confusing?

File: swift_scality_backend/diskfile.py [hunk 5]
```diff
@@ -354,12 +403,39 @@ def create(self, size=None):
...
+            return
+
+        # Keep the most recent Content-Type from either datafile or metafile metadata
+        if 'Content-Type' not in md_to_add and self._metadata is not None \
```
Comment (NicolasT) [inline swift_scality_backend/diskfile.py]: That's a lot of conditions in one `if`. I'd consider rewriting as
```python
if all([
        'Content-Type' not in md_to_add,
        self._metadata is not None,
        'Content-Type' in self._metadata['mf'],
        ...]):
    md_to_add(...)
    ...
```

File: test/unit/test_diskfile.py [hunk 6]
```diff
@@ -34,7 +34,6 @@
 from . import utils
 from .utils import make_client_collection
 
-
```
Comment (NicolasT) [inline test/unit/test_diskfile.py]: I believe this empty line is required for `flake` or some other checker.

File: tox.ini [hunk 7]
```diff
@@ -4,9 +4,10 @@
 # Newton: Swift 2.10.1
 # Ocata: Swift 2.13.1
 # Pike: Swift 2.15.1
+# Queens: Swift 2.17.0
 
 [tox]
-envlist = py27-swift{1.13.1,2.1.0,2.2.0,2.3.0,2.5.0,2.7.1,2.10.1,2.13.1,2.15.1}
+envlist = py27-swift{1.13.1,2.1.0,2.2.0,2.3.0,2.5.0,2.7.1,2.10.1,2.13.1,2.15.1,2.16.0,2.17.0}
```
Comment (NicolasT) [inline tox.ini]: Which OpenStack release brought Swift 2.16.0? Should be added to the list above.

File: swift_scality_backend/diskfile.py (hunk 2 above)
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: These are chosen by us, I'm open to alternative naming suggestions :)

File: swift_scality_backend/diskfile.py (hunk 3 above)
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: Yes, swift has a dependency on Python 2.7.
Also, the support of Python 2.6 was disabled in [this commit](https://github.com/scality/ScalitySproxydSwift/commit/0e03197de25c3b0e35267e561b69365260684caf)

File: swift_scality_backend/diskfile.py (hunk 4 above)
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: Right, fixed

File: swift_scality_backend/diskfile.py (hunk 5 above)
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: Fixed

File: test/unit/test_diskfile.py (hunk 6 above)
Comment (jgagnaire) [reply test/unit/test_diskfile.py]: No precise reason for the deletion of this empty line, I guess it is a leftover from a previous edit. Fixed now

File: tox.ini (hunk 7 above)
Comment (jgagnaire) [reply tox.ini]: 2.16.0 is Queens as well as 2.17.0, Fixed

File: swift_scality_backend/diskfile.py [hunk 8]
```diff
@@ -357,9 +406,38 @@ def create(self, size=None):
...
+            self.client_collection.get_write_client().put_meta(self._name, md_to_write)
+            return
+
+        # Keep the most recent Content-Type from either datafile or metafile metadata
+        if all(['Content-Type' not in md_to_add,
+                self._metadata is not None,
+                'Content-Type' in self._metadata['mf'],
+                self._metadata['mf']['Content-Type'] != self._metadata['df']['Content-Type'],
+                self._metadata['mf']['Content-Type-Timestamp'] >
+                self._metadata['df']['X-Timestamp']]):
```
Comment (jgagnaire) [inline swift_scality_backend/diskfile.py]: Restored the initial ugly if after testing: the all() syntax evaluates each of the conditions, without stopping at the first false.

This gives a KeyError on `self._metadata['mf']['Content-Type']` if the previous condition `'Content-Type' in self._metadata['mf']` is wrong

File: swift_scality_backend/diskfile.py (hunk 4 above)
Comment (jgagnaire) [reply swift_scality_backend/diskfile.py]: Ah right, missed this occurrence.. Fixed

**Pull request merged**